postgres/datadog_checks/postgres/statements.py

```diff
@@ -236,3 +227,50 @@ def row_keyfunc(row):
                 metrics.append((metric_name, value, tags))
 
         return metrics
+
+    def _normalize_query_stats(self, rows):
+        """
+        Given a list of rows, generate a new list of rows with an added normalized_query and query_signature
+        columns and a merged version of stats considering all queries mapping to a given normalized string
+        """
+        queries_by_normalized = {}
+        for row in rows:
+            row_with_normalization = dict(row)
+            try:
+                normalized_query = datadog_agent.obfuscate_sql(row['query'])
+                if not normalized_query:
+                    self.log.warning("Obfuscation of query '%s' resulted in empty query", row['query'])
+                    continue
+            except Exception as e:
+                # If query obfuscation fails, it is acceptable to log the raw query here because the
+                # pg_stat_statements table contains no parameters in the raw queries.
+                self.log.warning("Failed to obfuscate query '%s': %s", row['query'], e)
+                continue
+
+            row_with_normalization['normalized_query'] = normalized_query
+            row_with_normalization['query_signature'] = compute_sql_signature(normalized_query)
+            query_key = (
```

Comment:
It looks like we're duplicating the definition of the `row_keyfunc` here. Would it make more sense to move the "collapse rows by key" logic into `self._state.compute_derivative_rows`? It currently silently ignores duplicates but we could replace that silent logging with a "collapse rows by key" step so there are never any duplicate rows there. That will ensure the problem is solved for other databases if they have row collisions.

Reply:
I just realized that I was missing something here as well. The current `row_keyfunc` is not exactly a duplicate of this query_key mapping because it uses the postgres `queryid` but, now that I think about it, I think that's wrong and it should probably just use the computed `query_signature` as well.  

And yes, I think you're right about moving this to the `compute_derivative_rows`. That seems like the best way to do this (it's cleaner and potentially addresses other sources of conflicts).

Reply:
+1